### PR TITLE
add feature: export specifications in json format

### PR DESCRIPTION
**Pull request by Esorakouki**
add feature #205
add-json-export

### CODE REVIEW

**Comment by domoritz** [comment]
Thank you. We'll review this and get back to you soon.

**Comment by siddhant1** [comment]
@domoritz Have you testesed this ? This looks promising .
Should I test this once on my machine?

**Comment by domoritz** [comment]
This looks great! @siddhant1 that would be fantastic. 

We need to change something about the layout of these buttons soon. They should probably wrap instead of being one long row. But that's a separate PR.

**Comment by siddhant1** [comment]
Sure , that'll be a great improvement !

**Comment by domoritz** [comment]
One thing at a time.

**Comment by domoritz** [comment]
@siddhant1 Can you finish this so we can merge it?

**Comment by siddhant1** [comment]
Is there any change remaining in this one? 
I will test i today on my machine.
Also , should I make all the buttons in different rows?

**Comment by domoritz** [comment]
Rebase and test. I also don't think we need to change the width in css as done in this PR. Let's get this merged before making the export actions multiple rows. But it should be done, yes.

**Comment by siddhant1** [comment]
Tested ! Worked Great for me

File: src/components/header/renderer.tsx [hunk 1]
```diff
@@ -242,6 +249,28 @@ class Header extends React.Component<Props, State> {
     dlButton.classList.remove('disabled');
   }
```
Comment (domoritz) [inline src/components/header/renderer.tsx]: No need to be async

File: src/components/header/index.css [hunk 2]
```diff
@@ -399,7 +399,7 @@
 
 .export-button {
   flex: 1;
-  min-width: 280px;
+  min-width: 270px;
```
Comment (domoritz) [inline src/components/header/index.css]: Please undo

File: src/components/header/index.css [hunk 3]
```diff
@@ -411,7 +411,7 @@
   cursor: pointer;
   user-select: none;
   padding: 10px 20px;
-  margin-right: 40px;
+  margin-right: 30px;
 }
```
Comment (domoritz) [inline src/components/header/index.css]: Please undo

**Comment by domoritz** [comment]
I left a review. Please update and then we can merge

**Comment by domoritz** [comment]
We still use async.

**Comment by siddhant1** [comment]
Yeah , I had a doubt about it , I mentioned it above

**Comment by domoritz** [comment]
Where?

**Comment by siddhant1** [comment]
https://github.com/vega/editor/issues/237#issuecomment-470121959

**Comment by siddhant1** [comment]
https://github.com/vega/editor/pull/210#discussion_r263005659

**Comment by siddhant1** [comment]
So my doubt was , Doesn't this.props.view.ToImageURL returning promise which he is awaiting via await , I think thats why he made the function async .

File: src/components/header/renderer.tsx (hunk 1 above)
Comment (domoritz) [reply src/components/header/renderer.tsx]: Yes, but we don't need it for json.

Comment (siddhant1) [reply src/components/header/renderer.tsx]: Yes , I got it and pushed the changes too , this was a previous fork so there are some linting issues , I will fix them in my next pr.

File: src/components/header/renderer.tsx [hunk 4]
```diff
@@ -242,6 +249,28 @@ class Header extends React.Component<Props, State> {
...
+    if (
+      event.target &&
+      (event.target.matches('input') ||
+        event.target.matches('label') ||
+        event.target.matches('div.type-input-container'))
+    ) {
+      return;
+    }
+    const content = this.state.downloadVegaJSON ? this.props.vegaSpec : this.props.vegaLiteSpec;
+    const filename = this.state.downloadVegaJSON ? 'visualization.vg.json' : 'visualization.vl.json';
```
Comment (domoritz) [inline src/components/header/renderer.tsx]: In a follow up, please use template strings.

**Comment by domoritz** [comment]
Merge when test pass.

File: src/components/header/renderer.tsx (hunk 4 above)
Comment (siddhant1) [reply src/components/header/renderer.tsx]: Do they serve any benefit here?

Comment (domoritz) [reply src/components/header/renderer.tsx]: Yes, more readable code.

**Pull request merged**